tree/forest/v7/inc/ROOT/RTree.hxx

```diff
@@ -0,0 +1,148 @@
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RTree
+#define ROOT7_RTree
+
+#include <ROOT/RStringView.hxx>
+#include <ROOT/RTreeUtil.hxx>
+#include <ROOT/RTreeView.hxx>
+
+#include <memory>
+#include <utility>
+
+namespace ROOT {
+namespace Experimental {
+
+class RTreeEntry;
+class RTreeModel;
+
+namespace Detail {
+class RTreeSink;
+class RTreeSource;
+}
+
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::RTree
+\ingroup Forest
+\brief The RTree represents a live dataset, whose structure is defined by an RTreeModel
+
+RTree connects the static information of the RTreeModel to a source or sink on physical storage.
+Reading and writing requires use of the corresponding derived class RInputTree or ROutputTree.
+RTree writes only complete entries (rows of the data set).  The entry itself is not kept within the
+RTree, which allows for multiple concurrent entries for the same RTree.  Besides reading an entire entry,
+the RTree can expose tree views that read only specific branches.
+*/
+// clang-format on
+class RTree {
+private:
+   /// All trees that use the same model share its ownership
+   std::shared_ptr<RTreeModel> fModel;
+   /// The number of entries is constant for reading and reflects the current number of Fill() operations
+   /// when writing
+   TreeIndex_t fNentries;
+
+protected:
+   /// Only the derived RInputTree and ROutputTree can be instantiated
+   explicit RTree(std::shared_ptr<RTreeModel> model);
+
+public:
+   RTree(const RTree&) = delete;
+   RTree& operator =(const RTree&) = delete;
+   ~RTree();
+}; // RTree
+
+} // namespace Detail
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::RInputTree
+\ingroup Forest
+\brief An RTree that is used to read data from storage
+
+An input tree provides data from storage as C++ objects. The tree model can be created from the data on storage
+or it can be imposed by the user. The latter case allows users to read into a specialized tree model that covers
+only a subset of the branches in the tree. The tree model is used when reading complete entries.
+Individual branches can be read as well by instantiating a tree view.
+*/
+// clang-format on
+class RInputTree : public Detail::RTree {
+private:
+   std::unique_ptr<Detail::RTreeSource> fSource;
+
+public:
+   /// The user imposes a tree model, which must be compatible with the model found in the data on storage
+   RInputTree(std::shared_ptr<RTreeModel> model, std::unique_ptr<Detail::RTreeSource> source);
+   /// The model is generated from the tree metadata on storage
+   RInputTree(std::unique_ptr<Detail::RTreeSource> source);
+   ~RInputTree();
+
+   /// Analogous to Fill(), fills the default entry of the model
+   void GetEntry();
```

Comment:
The name is strange. It doesn't take any parameters and doesn't return anything, yet it *gets* something.

Reply:
The name would be familiar to those use to TTree::GetEntry.  Nonetheless it must return an error code or something as it *can* fail.

Reply:
ROOT is not consistent in the naming of this functionality.
`TTree` calls it `GetEntry`, `TTreeReader` calls it `SetEntry`.
I would argue that since the method does not _retrieve_ an entry, but it's job is to have side-effects, `SetEntry` is a better name (getters are typically side-effect free and return what they get).

Reply:
@bluehood Note that in the cases you mention that is indeed the difference (i.e. TTree::GetEntry will have the effect of reading the data from the file and placing it in the agreed upon memory location.  TTreeReader::SetEntry only set the cursor of which entry to read if the user access/dereference one of the readers.

Reply:
As far as users are concerned, there is no visible difference

Reply:
Depends on the users.   Some *are* interested in the timing of operations (i.e. what takes cpus, what requires network/disk access), etc.  eg.  if we switched to SetEntry, would you know if;
```
tree->SetEntry( entry );
branch1->SetEntry( entry );
branch2->SetEntry( entry );
branch3->SetEntry( entry );
```
is the redundance minorly bad or very bad?.

Reply:
There is some benefit is keeping the name for historical reason (i.e. users familiar with TTree::GetEntry will know what this is).

SetEntry is a setter and in the context of RForest is completely confusing (are you putting data there, setting cursor or what?) [i.e. I strongly oppose the name].

If GetEntry is not/no-longer an option, we should pick a name more expressive.   In the current implementation it seems that ReadEntry might be indicative of its functions.   If/when the reading is delayed we could use SetReadEntry or SetCurrentEntry or something that give an indication that you are setting a cursor and not the content

Cheers,
Philippe.

PS. > TTree::GetEntry() is not an entry getter: it doesn't even return the entry,

Yes it does get the entry content, it just returns it indirectly :) ["get" is here in the sense of 'grab', 'gather', 'fetch', 'obtain']